### PR TITLE
fixed #564 to execute all the UNIQUE INDEX DDLs before FK creation

### LINKED ISSUE #564: Import Schema dependency issue of Foreign Key on Unique Index
(This is observed during the migration from the source as PostgreSQL)

We get the following error while creating an FK during the import schema phase
```
 ERROR: there is no unique constraint matching given keys for referenced table "xyz" (SQLSTATE 42830)
```

SQL caused the above error:
```
    STATEMENT: ALTER TABLE ONLY public.abc ADD CONSTRAINT abc_fk1 FOREIGN KEY (c1) REFERENCES public.xyz(c1) ON DELETE CASCADE;
```

Error says there is no unique constraint on the`c1` column of table `xyz` so it can create a FK referencing that column.

But that UNIQUE constraint on that column is provided by the CREATE UNIQUE INDEX DDL:
```
CREATE UNIQUE INDEX xyz_ix2 ON public.xyz USING btree (c1);
```

Since we now postImportData phase, there is no way to create this index beforehand and on the other hand, all the FK are created before the postImportData phase. So FK creation is dependent on INDEX creation now, which voyager can’t handle today.

### PR DESCRIPTION
**Pull request by priyanshi-yb**
1. fixed #564 
2. Added a case in `migtests/tests/pg-tests/pg-constraints` for the same with validation.

### CODE REVIEW

File: yb-voyager/cmd/importSchema.go [hunk 1]
```diff
@@ -90,23 +90,40 @@ func importSchema() {
 	// Import ALTER TABLE statements from sequence.sql only after importing everything else
 	isAlterStatement := func(objType, stmt string) bool {
```
Comment (priyanshi-yb) [inline yb-voyager/cmd/importSchema.go:91]: I think we can change this `isAlterStatement` to `isSkipStatement` to make it general. Will do.

File: migtests/tests/pg-tests/pg-constraints/validate [hunk 2]
```diff
@@ -32,13 +34,22 @@ QUERIES_CHECK = {
 	'FORIEGN_CHECK': {
 		'query': "insert into foreign_test values (5,1,7);",
 		'code': 23503
-	}
+	},
+	'UNIQUE_INDEX': {
+		'query': "insert into xyz values(3);",
+		'code': 23505
+	},
+	'FORIEGN_CHECK_WITH_UNIQUE_INDEX': {
```
Comment (sanyamsinghal) [inline migtests/tests/pg-tests/pg-constraints/validate:42]: nit: `FOREIGN`

File: yb-voyager/cmd/importSchema.go [hunk 3]
```diff
@@ -90,23 +90,40 @@ func importSchema() {
...
 		stmt = strings.ToUpper(strings.TrimSpace(stmt))
-		return objType == "SEQUENCE" && strings.HasPrefix(stmt, "ALTER TABLE")
+		if objType == "SEQUENCE" && strings.HasPrefix(stmt, "ALTER TABLE"){
+			return true
+		} else if objType == "TABLE" && strings.Contains(stmt, "ALTER TABLE") && strings.Contains(stmt, "FOREIGN KEY"){
+			return true
+		} else if objType == "UNIQUE INDEX" && !strings.Contains(stmt, objType){
```
Comment (sanyamsinghal) [inline yb-voyager/cmd/importSchema.go:97]: are both of these if conditions at line - 97 and 99 required?
Or these can be merged into one?

File: yb-voyager/cmd/importSchema.go (hunk 1 above)
Comment (sanyamsinghal) [reply yb-voyager/cmd/importSchema.go:91]: or we can break it into multiple specific ones like `isALTERStatemet` , `createUniqueIndex` etc...

File: yb-voyager/cmd/importSchema.go (hunk 3 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/importSchema.go:97]: I am not sure if these can be merged, as one is to only run UNIQUE INDEX DDLs for objType UNIQUE INDEX and the other one is to skip only UNIQUE INDEX DDLs at the time of INDEX objType.

File: yb-voyager/cmd/importSchema.go (hunk 1 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/importSchema.go:91]: That is something I was also thinking but then as we pass only one skipFn to importSchemaInternal, we have to pass multiple functions to it I guess.

File: yb-voyager/cmd/importSchema.go [hunk 4]
```diff
@@ -90,21 +90,30 @@ func importSchema() {
...
+		return (objType == "SEQUENCE" || objType == "TABLE") && strings.HasPrefix(stmt, "ALTER TABLE")
 	}
 
 	skipFn := isAlterStatement
 	importSchemaInternal(exportDir, objectList, skipFn)
 
-	// Import the skipped ALTER TABLE statements from sequence.sql if it exists
-	if slices.Contains(objectList, "SEQUENCE") {
+	// Import the skipped ALTER TABLE statements from sequence.sql and table.sql if it exists
+	if slices.Contains(objectList, "SEQUENCE") || slices.Contains(objectList, "TABLE") {
```
[the remark this answers is not in the document]
Comment (priyanshi-yb) [reply yb-voyager/cmd/importSchema.go]: Done already

File: yb-voyager/cmd/importSchema.go (hunk 1 above)
Comment (sanyamsinghal) [reply yb-voyager/cmd/importSchema.go:91]: yeah, ok, maybe you can multiple conditions named separately and then add and pass them in one skipFn

File: yb-voyager/cmd/importSchema.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/cmd/importSchema.go]: yeah, it was an yesterday's comment which wasn't pushed

File: yb-voyager/cmd/importSchema.go (hunk 1 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/importSchema.go:91]: Yes, that can be done.

File: yb-voyager/cmd/importSchema.go [hunk 5]
```diff
@@ -88,25 +88,45 @@ func importSchema() {
...
+		return objType == "UNIQUE INDEX" && !strings.Contains(stmt, objType)
+	}
+	isUniqueIndexInIndex := func(objType, stmt string) bool{
+		return objType == "INDEX" && strings.Contains(stmt, "UNIQUE INDEX")
+	}
+	isSkipStatement := func(objType, stmt string) bool {
+		stmt = strings.ToUpper(strings.TrimSpace(stmt))
+		return isAlterTableInSequence(objType, stmt) || isAlterTableFKInTable(objType, stmt) || 
+			isNotUniqueIndexInUniqueIndex(objType, stmt) || isUniqueIndexInIndex(objType, stmt) 
+	}
```
[the remark this answers is not in the document]
Comment (priyanshi-yb) [reply yb-voyager/cmd/importSchema.go:107]: Okay, will make the change.

**Review by amit-yb — APPROVED** [review]
LGTM.

**Review by sanyamsinghal — APPROVED** [review]

**Pull request merged**